python/SWESimulators/Common.py

```diff
@@ -55,18 +54,24 @@ def __init__(self, cl_ctx, nx, ny, halo_x, halo_y, data, \
             self.nx_halo = nx + asymHalo[1] + asymHalo[3]
             self.ny_halo = ny + asymHalo[0] + asymHalo[2]
             
-        assert(host_data.shape[1] == self.nx_halo), str(host_data.shape[1]) + " vs " + str(self.nx_halo)
-        assert(host_data.shape[0] == self.ny_halo), str(host_data.shape[0]) + " vs " + str(self.ny_halo)
         
-        assert(data.shape == (self.ny_halo, self.nx_halo))
 
         #Upload data to the device
         mf = pyopencl.mem_flags
-        self.data = pyopencl.Buffer(cl_ctx, mf.READ_WRITE | mf.COPY_HOST_PTR, hostbuf=host_data)
-        self.holds_data = True
+        if (data is not None):
+            host_data = self.convert_to_float32(data)
+            self.bytes_per_float = host_data.itemsize
+            self.data = pyopencl.Buffer(cl_ctx, mf.READ_WRITE | mf.COPY_HOST_PTR, hostbuf=host_data)
+            
+            assert(host_data.shape[1] == self.nx_halo), str(host_data.shape[1]) + " vs " + str(self.nx_halo)
+            assert(host_data.shape[0] == self.ny_halo), str(host_data.shape[0]) + " vs " + str(self.ny_halo)
+            assert(self.bytes_per_float == 4)
+            assert(data.shape == (self.ny_halo, self.nx_halo))
+        else:
+            self.bytes_per_float = 4
+            self.data = pyopencl.Buffer(cl_ctx, mf.READ_WRITE, self.nx_halo*self.ny_halo*self.bytes_per_float)
         
-        self.bytes_per_float = host_data.itemsize
-        assert(self.bytes_per_float == 4)
+        self.holds_data = True
```

Comment:
Why is this here? What does it signify? Looks scary to me... Shouldn't these always own their own data?

Reply:
I think this is to avoid free'ing uninitialized memory on GPU. This was not added in this PR.

Reply:
That's right! I added self.holds_data to avoid freeing twice (which would give an exception)